### ISSUE TITLE
[Analyzer Plugin] ServerPlugin runs analyzeFile with different contexts for one path 

### ISSUE THREAD

**Issue by zltnDC** [dart-lang/sdk#49404]
I have encountered some problems with method from ServerPlugin:

```
@override
Future<void> analyzeFile(
    {required AnalysisContext analysisContext, required String path}) 
```
and AnalysisContext for several packages/directories.

```
analyzer_plugin: 0.11.0
```

I've prepared an example:

https://github.com/zltnDC/analyzer-plugin-bug

It seems that files in one directory are analyzed in different contexts and ignore contextRoot.includedPaths from AnalysisContext.

The analyzeFile method for one path is run several times with different contexts.

**Comment by scheglov** [comment]
This bug report is amazing!

I agree that it looks that `analyzeFile()` is invoked from `handleAffectedFiles()` for every `AnalysisContext`. Could you try to validate given `path` using `analysisContext.contextRoot.isAnalyzed(path)` inside `analyzeFile()`? If this works, I will put the same check into `analyzeFiles()` so that `analyzeFile()` invoked only for interesting analysis contexts.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/251442

**Comment by scheglov** [comment]
The CL landed, but I cannot publish `analyzer_plugin` yet, because it was updated to use not yet published changes to `package:analyzer`, and I'm not ready to publish them yet.

For now, you could keep the check in `analyzeFile`.

**Comment by scheglov** [comment]
I published `analyzer_plugin 0.11.1`, but you might also require `analyzer 4.3.1`.
